File: src/main/kotlin/top_150/_404_sum_of_left_leaves.java
package top_150;

public class _404_sum_of_left_leaves {
    public int sumOfLeftLeaves(TreeNode root) {
        return walk(root.left, true) + walk(root.right, false);
    }

    private int walk(TreeNode node, boolean isLeft) {
        if (node == null) return 0;
        if (node.left == null && node.right == null) {
            return isLeft ? node.val : 0;
        }

        var left = walk(node.left, true);
        var right = walk(node.right, false);
        return left + right;
    }


    public static void main(String[] args) {
        var s = new _404_sum_of_left_leaves();
        System.out.println(s.sumOfLeftLeaves(TreeNode.buildTree(3,9,20,null,null,15,7)));
    }
}

File: src/main/kotlin/easy/_404_Sum_of_Left_Leaves.kt
package easy

import common.TreeNode

fun sumOfLeftLeaves(root: TreeNode?): Int {
    root ?: return 0
    if (root.left == null && root.right == null) return 0
    val lefts = mutableListOf<Int>()
    fun sub(root: TreeNode?) {
        root ?: return
        val left = root.left
        val right = root.right
        if (left == null && right == null) return
        left?.let {
            if (it.left == null && it.right == null) {
                lefts.add(it.`val`)
            }
        }
        sub(left)
        sub(right)
    }
    sub(root)
    return lefts.sum()
}
